providers/keycloak_oidc.go

```diff
@@ -26,6 +26,18 @@ func NewKeycloakOIDCProvider(p *ProviderData) *KeycloakOIDCProvider {
 
 var _ Provider = (*KeycloakOIDCProvider)(nil)
 
+// CreateSessionFromToken converts Bearer IDTokens into sessions, extracting keycloak' roles
+func (p *KeycloakOIDCProvider) CreateSessionFromToken(ctx context.Context, token string) (*sessions.SessionState, error) {
+
+	ss, err := p.OIDCProvider.CreateSessionFromToken(ctx, token)
+	if err != nil {
+		return nil, err
+	}
+
+	return ss, p.extractRoles(ctx, ss)
```

Comment:
Even though I wrote this provider, I'm not a Keycloak user so I had to rely on the community (so please excuse any misunderstandings on my part 😄 )

I thought all this was on the AccessToken -- when I look below at `extractRoles` it is operating on the AccessToken, not the IDToken.

`CreateSessionFromToken` is intended to consume an IDToken based in via the Bearer header.

If Keycloak IDTokens have these fields now (and not the Access Token), it seems like we have a bigger issue and need to align how we consume data from the tokens?

Reply:
On the contrary, it probably was me under the incorrect assumption that the Bearer token should be an access token, given that there's a specific (draft) profile for them (rfc9068), and rfc6750 seems to mention Access Token to be used as bearer, but never mentions IDTokens. You're right about Keycloaks not including those fields by default, however it's quite straightforward to make idtokens include the resource_access structure using a predefined mapper. Token usage here is a bit fuzzy, this might work with both access token and IDToken, specially for keycloak, that uses JWT tokens for everything.

Reply:
@NickMeves is there anything I could do to improve the change? In its current implementation, the keycloak provider won't allow combining role authorization with --skip-jwt-bearer-tokens option, very useful if both users and machines are accessing the same service.